subsys/ipc/ipc_service/backends/ipc_rpmsg_static_vrings.c

```diff
@@ -0,0 +1,400 @@
+/*
+ * Copyright (c) 2021 Carlo Caione <ccaione@baylibre.com>
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <logging/log.h>
+#include <zephyr.h>
+#include <cache.h>
+#include <device.h>
+#include <sys/atomic.h>
+
+#include <ipc/ipc_service_backend.h>
+#include <ipc/ipc_static_vrings.h>
+#include <ipc/ipc_rpmsg.h>
+
+#include <drivers/mbox.h>
+
+#include "ipc_rpmsg_static_vrings.h"
+
+#define DT_DRV_COMPAT	subsys_ipc_openamp_static_vrings
+
+#define WQ_PRIORITY	(0)
+#define WQ_STACK_SIZE	CONFIG_IPC_SERVICE_BACKEND_RPMSG_WQ_STACK_SIZE
+
+K_KERNEL_STACK_DEFINE(mbox_stack, WQ_STACK_SIZE);
+static struct k_work_q mbox_wq;
+
+struct backend_data_t {
+	/* RPMsg */
+	struct ipc_rpmsg_instance rpmsg_inst;
+
+	/* Static VRINGs */
+	struct ipc_static_vrings vr;
+
+	/* General */
+	struct k_work mbox_work;
+	unsigned int role;
+	atomic_t inited;
+};
+
+struct backend_config_t {
+	unsigned int role;
+	uintptr_t shm_addr;
+	size_t shm_size;
+	struct mbox_channel mbox_tx;
+	struct mbox_channel mbox_rx;
+};
+
+static void rpmsg_service_unbind(struct rpmsg_endpoint *ep)
+{
+	rpmsg_destroy_ept(ep);
+}
+
+static struct ipc_rpmsg_ept *get_ept_slot_with_name(struct ipc_rpmsg_instance *rpmsg_inst,
+						    const char *name)
+{
+	struct ipc_rpmsg_ept *rpmsg_ept;
+
+	for (size_t i = 0; i < NUM_ENDPOINTS; i++) {
+		rpmsg_ept = &rpmsg_inst->endpoint[i];
+
+		if (strcmp(name, rpmsg_ept->name) == 0) {
+			return &rpmsg_inst->endpoint[i];
+		}
+	}
+
+	return NULL;
+}
+
+static struct ipc_rpmsg_ept *get_available_ept_slot(struct ipc_rpmsg_instance *rpmsg_inst)
+{
+	return get_ept_slot_with_name(rpmsg_inst, "");
```

Comment:
right. We should probably just forbid endpoints with an empty name.

Reply:
Yes, it is. The goal of this function is picking up an empty endpoint slot to fill. An empty endpoint slot is recognized by having an empty name.

In general it works this way: if the name is empty then the endpoint is available and it can be filled in two cases:
1. When the remote endpoint is advertising a new endpoint that it is still unknown to the host, see: https://github.com/zephyrproject-rtos/zephyr/blob/32812e7b62e1fb88c471f773332a8f274b115e6e/subsys/ipc/ipc_service/backends/ipc_rpmsg_static_vrings.c#L159-L166
2. When the host is registering a new endpoint that it is still not advertised by the remote endpoint. In this case we cache the name and wait for the advertising from the remote, see:
https://github.com/zephyrproject-rtos/zephyr/blob/32812e7b62e1fb88c471f773332a8f274b115e6e/subsys/ipc/ipc_service/backends/ipc_rpmsg_static_vrings.c#L325-L329

Reply:
> Yes, it is. The goal of this function is picking up an empty endpoint slot to fill. An empty endpoint slot is recognized by having an empty name.
Understood now , thanks for the clarification 

> 
> In general it works this way: if the name is empty then the endpoint is available and it can be filled in two cases:
> 
> 1. When the remote endpoint is advertising a new endpoint that it is still unknown to the host, see: https://github.com/zephyrproject-rtos/zephyr/blob/32812e7b62e1fb88c471f773332a8f274b115e6e/subsys/ipc/ipc_service/backends/ipc_rpmsg_static_vrings.c#L159-L166

`name == null` should not occur on a name service annoucement. What about treat it as an error?


> 2. When the host is registering a new endpoint that it is still not advertised by the remote endpoint. In this case we cache the name and wait for the advertising from the remote, see:
>    https://github.com/zephyrproject-rtos/zephyr/blob/32812e7b62e1fb88c471f773332a8f274b115e6e/subsys/ipc/ipc_service/backends/ipc_rpmsg_static_vrings.c#L325-L329

Same here `cfg->name` can not be null as tested before in the function

Reply:
ACK